src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,551 @@
+			exit();
+		}
+		$this->handler_args['handler_error'] = 'RAU Shutdown Function';
+		$this->handler_args['error_msg']     = $last_error['message'];
+		$this->handler();
+	}
+
+	/**
+	 * Check for errors only caused by an active plugin using 'include()'.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $error_msg Error message from handler.
+	 * @return array|bool
+	 */
+	private function check_passing_errors( $error_msg ) {
+		if ( empty( $error_msg ) ) {
+			return array();
+		}
+		preg_match( '/(' . implode( '|', static::$error_exceptions ) . ')/', $error_msg, $matches );
+		if ( ! empty( $matches ) ) {
+			return $this->handler_args['result'];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Handles errors by running Rollback.
+	 *
+	 * @since 6.4.0
+	 */
+	private function handler() {
+		error_log( var_export( $this->handler_args['handler_error'] . ' - ' . $this->handler_args['error_msg'], true ) );
+		if ( in_array( $this->handler_args['hook_extra']['plugin'], self::$fatals, true ) ) {
+			return;
+		}
+		self::$fatals[] = $this->handler_args['hook_extra']['plugin'];
+
+		$this->cron_rollback();
+
+		/*
+		 * This possibly helps to avoid a potential race condition on servers that may start to
+		 * process the next plugin for auto-updating before the handler can pick up an error from
+		 * the previously processed plugin.
+		 */
+		sleep( 2 );
+
+		/*
+		 * If a plugin upgrade fails prior to a theme upgrade running, the plugin upgrader will have
+		 * hooked the 'Plugin_Upgrader::delete_old_plugin()' method to 'upgrader_clear_destination',
+		 * which will return a `WP_Error` object and prevent the process from continuing.
+		 *
+		 * To resolve this, the hook must be removed using the original plugin upgrader instance.
+		 */
+		remove_filter( 'upgrader_clear_destination', array( static::$plugin_upgrader, 'delete_old_plugin' ) );
+
+		$this->restart_updates_and_send_email();
+	}
+
+	/**
+	 * Rolls back during cron.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+	 */
+	private function cron_rollback() {
+		global $wp_filesystem;
+
+		$temp_backup = array(
+			'temp_backup' => array(
+				'dir'  => 'plugins',
+				'slug' => dirname( $this->handler_args['hook_extra']['plugin'] ),
+				'src'  => $wp_filesystem->wp_plugins_dir(),
+			),
+		);
+
+		$rollback_updater = new WP_Upgrader();
+
+		// Set private $temp_restores variable.
+		$ref_temp_restores = new ReflectionProperty( $rollback_updater, 'temp_restores' );
+		$ref_temp_restores->setAccessible( true );
+		$ref_temp_restores->setValue( $rollback_updater, $temp_backup );
+
+		// Set private $temp_backups variable.
+		$ref_temp_backups = new ReflectionProperty( $rollback_updater, 'temp_backups' );
+		$ref_temp_backups->setAccessible( true );
+		$ref_temp_backups->setValue( $rollback_updater, $temp_backup );
+
+		// Call Rollback's restore_temp_backup().
+		$rollback_updater->restore_temp_backup();
```

Comment:
> Or if it fails, is the assumption that's the best it can do?

That's essentially it yeah.

During an upgrade, restoration and deletion of temp backups occurs during the `'shutdown'` action, triggered in a registered shutdown function, when the upgrade _process_ fails.

While these methods _might_ fail, by this point, the previously installed version has already been backed up, the new version downloaded and unpacked to the plugins directory. We should therefore have the permissions and diskspace to do a restore and delete. PHP's `shutdown` is immune to timeouts, so that's another potential cause of failure avoided.

Nevertheless, we worded the email carefully in saying "should have been rolled back" to account for cases that are very hard to predict given our established permissions, diskspace, and timeout immunity.

Reply:
Thank you! That makes sense. I would love to see this described in a comment, since it may be clear to some folks, but wasn't immediately so for me, and might make it easier for folks to work on in the future.